Add doc dir for Keptn 0.16 (#1228)
keptn/keptn.github.io: content/docs/0.16.x/operate/backup_and_restore/index.md

---
title: Back up and restore Keptn
description: How to back up and restore Keptn.
weight: 50
keywords: backup
---

> To secure all your data in your projects' Git repository, as well as all events that have occurred for these projects, you need to 

Reviewer: ```suggestion
To secure all your data in your project's Git repository, as well as all events that have occurred for these projects, you must
```

> back up the data within the *Configuration Service*, the *MongoDB*, and credentials to your projects' Git upstream repos (if you have configured those).

Reviewer: ```suggestion
back up the data within the *Configuration Service*, the *MongoDB*, and credentials to your project's Git upstream repos (if you have configured those).
```

This should include all projects you have onboarded with Keptn and a `lost+found` directory, which can be ignored.

```console
$ ls config-svc-backup/config
lost+found	sockshop my-project
```

### Back up Mongodb Data

The next step is to create a backup of the mongodb database in the `keptn-datastore` namespace using the `mongodump` tool.`

Create a directory for the MongoDB backup:

```console
mkdir mongodb-backup
chmod o+w mongodb-backup
```

Copy the data from the MongoDB to your local directory:

```console
MONGODB_ROOT_USER=$(kubectl get secret mongodb-credentials -n keptn -ojsonpath={.data.mongodb-root-user} | base64 -d)
MONGODB_ROOT_PASSWORD=$(kubectl get secret mongodb-credentials -n keptn -ojsonpath={.data.mongodb-root-password} | base64 --decode)

kubectl exec svc/keptn-mongo -n keptn -- mongodump --authenticationDatabase admin --username $MONGODB_ROOT_USER --password $MONGODB_ROOT_PASSWORD -d keptn -h localhost --out=/tmp/dump

MONGODB_POD=$(kubectl get pods -n keptn -lapp.kubernetes.io/name=mongo -ojsonpath='{.items[0].metadata.name}')
kubectl cp keptn/$MONGODB_POD:/tmp/dump ./mongodb-backup/ -c mongodb
```

### Back up Git credentials

> For projects where you have configured a Git upstream repository, we advise you to back up the Kubernetes secrets needed to access those.

Reviewer: ```suggestion
For projects where you have configured a Git upstream repository, you should back up the Kubernetes secrets needed to access those.
```

kubectl cp ./reset-git-repos.sh keptn/$CONFIG_SERVICE_POD:/data/config -c configuration-service
kubectl exec -n keptn $CONFIG_SERVICE_POD -c configuration-service -- chmod +x -R ./data/config/reset-git-repos.sh
kubectl exec -n keptn $CONFIG_SERVICE_POD -c configuration-service -- ./data/config/reset-git-repos.sh
``` 

### Restore MongoDB data

Copy the content of the mongodb-backup directory you have created earlier into the pod running the MongoDB:

```console
MONGODB_POD=$(kubectl get pods -n keptn -lapp.kubernetes.io/name=mongo -ojsonpath='{.items[0].metadata.name}')
kubectl cp ./mongodb-backup/ keptn/$MONGODB_POD:/tmp/dump -c mongodb
```

Import the MongoDB dump into the database using the following command:

```console
MONGODB_ROOT_USER=$(kubectl get secret mongodb-credentials -n keptn -ojsonpath={.data.mongodb-root-user} | base64 -d)
MONGODB_ROOT_PASSWORD=$(kubectl get secret mongodb-credentials -n keptn -ojsonpath={.data.mongodb-root-password} | base64 -d)
kubectl exec svc/keptn-mongo -n keptn -- mongorestore --drop --preserveUUID --authenticationDatabase admin --username $MONGODB_ROOT_USER --password $MONGODB_ROOT_PASSWORD /tmp/dump
```

### Restore Git credentials

> To re-establish the communication between the configuration-service and your projects' upstream repositories, restore the K8s secrets you have previously stored in the `keptn-backup` directory , using `kubectl apply`:

Reviewer: ```suggestion
To re-establish the communication between the configuration-service and your projects' upstream repositories, restore the Kubernetes secrets you have previously stored in the `keptn-backup` directory , using `kubectl apply`:
```
